activerecord/lib/active_record/associations/lazy_preloader.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module Associations
+    class LazyPreloader
+      # Implements the details of lazy eager loading of Active Record associations.
+      #
+      # Suppose that you have the following two Active Record models:
+      #
+      #   class Author < ActiveRecord::Base
+      #     # columns: name, age
+      #     has_many :books
+      #   end
+      #
+      #   class Book < ActiveRecord::Base
+      #     # columns: title, sales, author_id
+      #   end
+      #
+      # When you load an author with all associated books Active Record will make
+      # only one query:
+      #
+      #   authors = Author.lazy_preload(:books).where(name: ['bell hooks', 'Homer']).to_a
+      #
+      #   => SELECT `authors`.* FROM `authors` WHERE `name` IN ('bell hooks', 'Homer')
+      #
+      # Then you probably will want to retrieve books while iterating authors
+      #
+      #   authors.each { |author| read author.books }
+      #
+      # And only after books method invocation second query will be executed
+      # to load all books for authors in the collection:
+      #
+      # => SELECT `books`.* FROM `books` WHERE `author_id` IN (2, 5)
+      #
+      class WeakRegistry
```

Comment:
There is no longer a _weak_ registry, isn't it? This naming is a little bit confusing.

What about calling the methods `store` and `fetch`, and probably change the behaviour of the latter one to:

```ruby
def fetch
  if record.instance_variable_defined?(:@_lazy_preloader)
    # here we can assume that if the variable exists then it's not nil
    # so we don't have to nil-check later
    yield record.instance_variable_get :@_lazy_preloader
  end
end

# and the use it like this
def lazy_preload
  LazyPreloader::Registry.fetch(owner) do |preloader|
    preloader.preload reflection.name if preloader.should_load?(reflection.name)
  end
end
```

Reply:
You are correct, thank you for your comments!

Reply:
Actually mistake was in naming `Registry` `WeakRegistry`, because `weak` is kind of low-level implementation detail, which is unnecessary to know for users of registry.